src/libmongoc/src/mongoc/mongoc-secure-channel.c

```diff
@@ -202,21 +266,25 @@ mongoc_secure_channel_setup_certificate_from_file (const char *filename)
                              &hKey);           /* phKey, OUT */
    if (!success) {
       MONGOC_ERROR ("CryptImportKey for private key failed with error 0x%.8X", (unsigned int) GetLastError ());
+      CryptReleaseContext (provider, 0);
       goto fail;
    }
+   CryptDestroyKey (hKey);
 
    /* https://msdn.microsoft.com/en-us/library/windows/desktop/aa376573%28v=vs.85%29.aspx
     */
+   // CertSetCertificateContextProperty takes ownership of `provider`.
    success = CertSetCertificateContextProperty (cert,                         /* pCertContext */
                                                 CERT_KEY_PROV_HANDLE_PROP_ID, /* dwPropId */
                                                 0,                            /* dwFlags */
                                                 (const void *) provider);     /* pvData */
-   if (success) {
-      TRACE ("%s", "Successfully loaded client certificate");
-      return cert;
+   if (!success) {
+      MONGOC_ERROR ("Can't associate private key with public key: 0x%.8X", (unsigned int) GetLastError ());
+      goto fail;
    }
 
-   MONGOC_ERROR ("Can't associate private key with public key: 0x%.8X", (unsigned int) GetLastError ());
+   TRACE ("%s", "Successfully loaded client certificate");
+   ret = true;
 
 fail:
    SecureZeroMemory (pem, pem_length);
```

Comment:
```c
   if (pem) {
      SecureZeroMemory (pem, pem_length);
      bson_free (pem);
   }
```
If `read_file_and_null_terminate` fails and returns `NULL`, this call to `SecureZeroMemory` is invalid (it does not accept a null pointer according to its `_In_` annotation).
```
entering windows exception handler
exception '(access violation)' at '0x00007FF6A2010AE3', terminating
begin stack trace
test-libmongoc : RtlSecureZeroMemory
test-libmongoc : mongoc_secure_channel_setup_certificate_from_file
```